[DOC 364] lopp/doc-364-write-creating-dagster-code-locations (#23967)
dagster-io/dagster: docs/docs-beta/docs/dagster-plus/deployment/code-locations.md

1. You should have a Dagster project, we recommend setting this project up as a Python module following our [recommended project structure](/tutorial/create-new-project).

2. You should know whether or not you are using Dagster Plus Serverless or Dagster Plus Hybrid. Your administrator will know that, or you can find the agent type by navigating to your Dagster+ organization, clicking on the "Deployment" page, and selecting the "Agent" tab. Dagster+ Serverless organizations will have a label "Managed by Dagster+".

</details>

Adding a code location follows two general steps:
- Putting your Dagster code in a place that's accessible by your agent. For Dagster+ Serverless you can skip this step. For Dagster+ Hybrid this usually means building a Docker image with your code and pushing it to a registry.
- Notifying Dagster+ of the new or updated code location. This will be done by using the Dagster+ Python client.

Often these two steps are handled by CICD connected to your Git repository.

## Add a new code location manually

Start by installing the `dagster-cloud` Python client:

```
pip install dagster-cloud
```

Next you will want need to authenticate this Python client:

> 1. Navigate to your Dagster+ organization. Click on your user icon in the upper right corner, then select "Organization settings". Select "Tokens" and press "Create user token". Copy the token.

Reviewer: This is more than one step - let's break it up for readability:

```suggestion
1. In the Dagster+ UI, click on your user icon in the upper right corner.
2. Click **Organization settings**, then the **Tokens** tab.
3. Click the **Create user token** button. 
4. Copy the token.
```

2. Set the following environment variables:

> ```bash
> export  DAGSTER_CLOUD_ORGANIZATION="organization-name" # if your URL is https://acme.dagster.plus your organization name is "acme"
> export  DAGSTER_CLOUD_API_TOKEN="your-token"
> ```

Reviewer: This needs to be indented to correctly nest under this list item:

````suggestion
   ```bash
   export  DAGSTER_CLOUD_ORGANIZATION="organization-name" # if your URL is https://acme.dagster.plus your organization name is "acme"
   export  DAGSTER_CLOUD_API_TOKEN="your-token"
   ```
````

I'm not going to add comments for all other instances of this, but it's something to check for when writing.

3. Add the code location. The following example assumes you are running the command from the top-level working directory of your Dagster project with a project named "quickstart" structured as a Python module named "quickstart".

```bash
$pwd
/projects/quickstart
$tree
/quickstart
    setup.py
    pyproject.toml
    /quickstart
        __init__.py
        /assets
        /resources
```

If you are using Dagster+ Serverless, you'll want to run:

```bash

> dagster-cloud serverless deploy-python-executable --deployment prod --location-name quickstart --module-name quickstart

Reviewer 2: Would location-name and module-name ever be different?

```

> Use `dagster-cloud serverless deploy-python-executable --help` to see all available options. Running the command multiple times with the same location name will *update* the code location. Running the command with a new location name will *add* a code location.

Reviewer: Do we need to tell them how to view CLI help?

> If you are using Dagster+ Hybrid, make sure you have deployed the code appropriately by building a Docker image and pushing it to a Git repository. Then run:

Reviewer 2: Is there a guide on doing this that we should write? If so, link to it via /todo link.

> If you are using Dagster+ Hybrid, make sure you have deployed the code appropriately by building a Docker image and pushing it to a Git repository. Then run:

Reviewer 2: What is `--image` how do I know what to put there?

```bash
dagster-cloud deployment add-location --deployment prod --location-name quickstart --module-name quickstart --image 764506304434.dkr.ecr.us-west-2.amazonaws.com/hooli-data-science-prod:latest

> ```

Reviewer: I think putting these instructions into tabs will be helpful in reducing noise for folks who only care about one deployment type. Let me know if you want to take a swing, otherwise I'll do it on a follow up.

Reviewer: Additionally, how should the user confirm that everything worked correctly?

Reviewer 2: I wonder if there's a way to make all the tabs respond to a single click (so if you click Hybrid..they all become Hybrid. I'm sure the answer is yes)

Reviewer 2: ah cool. I'm positive we can create a toggle on the docs somewhere that sets this value throughout the site. Maybe file a ticket for post-sprint?

> ## Adding a code location in a new Git repository

Reviewer: This section describes what the user needs to have to do this, but not actually how they do it.

> Adding a code location to a Git repository follows the same steps as adding a code location manually, but automates those steps by running them through CICD. The best way to get started is to structure your repository following one of the Dagster+ template repositories:

Reviewer: ```suggestion
Adding a code location to a Git repository follows the same steps as adding a code location manually, but automates those steps by running them through CI/CD. 

The best way to get started is to structure your repository by following one of the Dagster+ template repositories as an example:
```

I added what I thought was a missing word to that second sentence, but looking at it again, I'm not sure this is actually right. Should it be something like "... by using one of the following Dagster+ templates to create a new repository"? I interpreted "following one of" to mean "I should create a repo on my own that follows the structure of these repositories here"

- [GitHub repository with Dagster+ Serverless](https://github.com/dagster-io/dagster-cloud-serverless-quickstart/)
- [GitHub repository with Dagster+ Hybrid](https://github.com/dagster-io/dagster-cloud-hybrid-quickstart/)
- [Gitlab CICD for Dagster+ Serverless](https://github.com/dagster-io/dagster-cloud-action/blob/main/gitlab/serverless-ci.yml)
- [Gitlab CICD for Dagster+ Hybrid](https://github.com/dagster-io/dagster-cloud-action/blob/main/gitlab/hybrid-ci.yml)

Overall, the Git repository should contain:

1. Your Dagster code, structured as a Python module. The repository might look like this:

```bash
README.md
dagster_cloud.yaml
Dockerfile
/.github
    /workflows
        dagster-cloud-deploy.yml
setup.py
pyproject.toml
/quickstart
    __init__.py
    definitions.py
    /assets
        ...
    /resources
        ...
```

> 2. A `dagster_cloud.yaml` file with the settings for your code location. Here is an example:

Reviewer 2: Link to a `dagster_cloud.yaml` reference file here please with a todo

> ```yaml

Reviewer: ```suggestion
```yaml title="dagster_cloud.yaml"
```

locations:
  - location_name: quickstart
    code_source:
      package_name: quickstart
```

> 3. A CICD workflow file that contains the steps for adding your code location. Here is an example workflow file for a Dagster+ Hybrid organization:

Reviewer: ```suggestion
3. A CI/CD workflow file that contains the steps for adding your code location. Here is an example workflow file for a Dagster+ Hybrid organization:
```

> 3. A CICD workflow file that contains the steps for adding your code location. Here is an example workflow file for a Dagster+ Hybrid organization:

Reviewer 2: Is this a Github Action file? Where did this come from?

stages:
  - build
  - deploy

build:
  stage: build
  image: docker:latest
  services:
    - docker:dind
  before_script:
    # # For Gitlab Container Registry
    # - echo $CI_JOB_TOKEN | docker login --username $CI_REGISTRY_USER --password-stdin $REGISTRY_URL
    # # For DockerHub
    # - echo $DOCKERHUB_TOKEN | docker login --username $DOCKERHUB_USERNAME --password-stdin $REGISTRY_URL
    # # For AWS Elastic Container Registry (ECR)
    # - apk add --no-cache curl jq python3 py3-pip
    # - pip install awscli
    # - echo $AWS_ECR_PASSWORD | docker login --username AWS --password-stdin $IMAGE_REGISTRY
    # # For Google Container Registry (GCR)
    # - echo $GCR_JSON_KEY | docker login --username _json_key --password-stdin $REGISTRY_URL
  script:
    - docker build . -t $IMAGE_REGISTRY:$IMAGE_TAG
    - docker push $IMAGE_REGISTRY:$IMAGE_TAG

deploy:
  stage: deploy
  dependencies:
    - build
  image: ghcr.io/dagster-io/dagster-cloud-action:0.1.43
  script:
    - dagster-cloud deployment add-location --deployment prod --image
    $IMAGE_REGISTRY:$IMAGE_TAG --location-name quickstart --package-name quickstart
```

## Adding a new code location to a Git monorepo

> Many organizations use a Git monorepo to contain multiple Dagster projects. Here is an example of DagsterLab's own [internal data engineering Git repository](https://github.com/dagster-io/dagster-open-platform). To add a new location to a monorepo, you will create a new directory containing:

Reviewer: ```suggestion
Many organizations use a Git monorepo to contain multiple Dagster projects. Here is an example of DagsterLab's own [internal data engineering Git repository](https://github.com/dagster-io/dagster-open-platform).

To add a new code location to a monorepo, create a new directory that contains:
```

- Suggesting we add `code` to `location` to make it specific and consistent
- Also that we remove `you will` and instead just use understood you to make this a command
- Adding a line break to visually break up the copy

- Your Dagster project
- For Dagster+ Hybrid, you will typically add a Dockerfile:

```
FROM python:3.11-slim
# Add any steps to install project system dependencies like java

WORKDIR /opt/dagster/app

COPY . /opt/dagster/app

# Add steps to install the Python dependencies for your Dagster project
# into the default Python on PATH
# For example, this project uses setup.py and we install all dependencies into the Docker container
# using `pip`.

RUN pip install -e .
```

The final repository structure might look like this:

```
README.md
dagster_cloud.yaml
/.github
    ...
/shared
    setup.py
    pyproject.toml
    /shared
        __init__.py
        utilities.py
/core
    setup.py
    pyproject.toml
    Dockerfile
    /core
        definitions.py
        __init__.py
        /assets
/new-code-location
    setup.py
    pyproject.toml
    Dockerfile
    /new-code-location
        definitions.py
        __init__.py
        /assets
```

> Then you will update the `dagster_cloud.yaml` file in the root of the Git repository:

Reviewer: Nit - remove `you will` and instead have this be `Then, update the...`

What needs to be updated? Should this instead be `Update the dagster_cloud.yaml with your registry image`?

```yaml
locations:
  - location_name: core
    code_source:
      package_name: core
    build:
      directory: ./core
      registry:  your-registry/image # eg 764506304434.dkr.ecr.us-west-2.amazonaws.com/core
  - location_name: new
    code_source:
      package_name: new
    build:
      directory: ./new
      registry: your-registry/image # eg 764506304434.dkr.ecr.us-west-2.amazonaws.com/new
```

> The monorepo should have CICD configured to deploy your changes and add or update your new code location.

Reviewer: How can they confirm this worked?
